adam_core/dynamics/propagation.py

```diff
@@ -0,0 +1,174 @@
+        Numerical tolerance to which to compute universal anomaly using the Newtown-Raphson
+        method.
+
+    Returns
+    -------
+    orbits : `~jax.numpy.ndarray` (N, 6)
+        Orbit propagated to each MJD with position in units of au and velocity in units
+        of au per day.
+    """
+    r = orbit[0:3]
+    v = orbit[3:6]
+    dt = t1 - t0
+
+    lagrange_coeffs, stumpff_coeffs, chi = calc_lagrange_coefficients(
+        r, v, dt, mu=mu, max_iter=max_iter, tol=tol
+    )
+    r_new, v_new = apply_lagrange_coefficients(r, v, *lagrange_coeffs)
+
+    return jnp.array([r_new[0], r_new[1], r_new[2], v_new[0], v_new[1], v_new[2]])
+
+
+# Vectorization Map: _propagate_2body
+_propagate_2body_vmap = jit(
+    vmap(_propagate_2body, in_axes=(0, 0, 0, None, None, None), out_axes=(0))
+)
+
+
+def propagate_2body(
+    orbits: Orbits,
+    times: Time,
+    mu: float = MU,
+    max_iter: int = 1000,
+    tol: float = 1e-14,
+) -> Orbits:
+    """
+    Propagate orbits using the 2-body universal anomaly formalism.
+
+    Parameters
+    ----------
+    orbits : `~jax.numpy.ndarray` (N, 6)
+        Cartesian orbits with position in units of au and velocity in units of au per day.
+    times : `~astropy.time.core.Time` (M)
+        Epochs to which to propagate each orbit. If a single epoch is given, all orbits are propagated to this
+        epoch. If multiple epochs are given, then each orbit to will be propagated to each epoch.
+    mu : float, optional
+        Gravitational parameter (GM) of the attracting body in units of
+        au**3 / d**2.
+    max_iter : int, optional
+        Maximum number of iterations over which to converge. If number of iterations is
+        exceeded, will return the value of the universal anomaly at the last iteration.
+    tol : float, optional
+        Numerical tolerance to which to compute universal anomaly using the Newtown-Raphson
+        method.
+
+    Returns
+    -------
+    orbits : `~adam_core.orbits.orbits.Orbits` (N*M)
+        Orbits propagated to each MJD.
+    """
+    # Lets extract the cartesian orbits and times from the orbits object
+    cartesian_orbits = orbits.coordinates.values
+    t0 = orbits.coordinates.time.to_astropy().tdb.mjd
+    t1 = times.tdb.mjd
+    orbit_ids = orbits.orbit_id.to_numpy(zero_copy_only=False)
+    object_ids = orbits.object_id.to_numpy(zero_copy_only=False)
+
+    # Lets stack the orbits into a single array shaped by the number of orbits and number of times
+    # and then pass this to the vectorized map version of _propagate_2body
+    n_orbits = cartesian_orbits.shape[0]
+    n_times = len(times)
+    orbit_ids_ = np.repeat(orbit_ids, n_times)
+    object_ids_ = np.repeat(object_ids, n_times)
+    orbits_array_ = np.repeat(cartesian_orbits, n_times, axis=0)
+    t0_ = np.repeat(t0, n_times)
+    t1_ = np.tile(t1, n_orbits)
+
+    orbits_propagated = _propagate_2body_vmap(
+        orbits_array_, t0_, t1_, mu, max_iter, tol
+    )
```

Comment:
It strikes me that, for the same orbit, and the same t0, we worth through a sequence of `t1_` values - it's the only parameter that is varying. If you chase that through the code, the _only instruction_ that is different is https://github.com/B612-Asteroid-Institute/adam_core/blob/633f818f0242a89211a7ff12125fb1f7acb2d808/adam_core/dynamics/chi.py#L68. So, _in theory_, that is the first point where vectorization ought to occur; all the operations before there are identical across the orbit's time steps.

Could make the code marvelously complicated to accommodate that, though...